File: 이주은/Heap/prob3.java
//[프로그래머스] 이중우선순위큐(https://school.programmers.co.kr/learn/courses/30/lessons/42628)

import java.util.*;

class Solution {
    public int[] solution(String[] operations) {
        int[] answer = {0, 0};
        PriorityQueue<Integer> minHeap = new PriorityQueue<>();
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>(Collections.reverseOrder());
        
        for(String operation : operations){
            String[] tokens = operation.split(" ");
            int num;
            if(tokens[0].equals("I")){
                num = Integer.parseInt(tokens[1]);
                minHeap.add(num);
                maxHeap.add(num);
            }
            else if(!minHeap.isEmpty()){
                if(tokens[1].equals("1")){
                    num = maxHeap.poll();
                    minHeap.remove(num);
                }  
                else
                    num = minHeap.poll();
                    maxHeap.remove(num);
            }
        }
        if(!minHeap.isEmpty()){
            answer[0] = maxHeap.poll();
            answer[1] = minHeap.poll();
        }
        return answer;
    }
}

File: 전예진/Heap/prob3.py
import heapq

def solution(operations):
    answer = []
    
    max_heap = []
    min_heap = []
    visited = [False] * 1000001
    
    for i in range(len(operations)):
        operation, number = operations[i].split()
        number = int(number)
        
        if operation == 'I':
            heapq.heappush(max_heap, (-number, i ))
            heapq.heappush(min_heap, (number, i))
            visited[i] = True

        elif operation == 'D' and number == 1:
            while max_heap and not visited[max_heap[0][1]]:
                heapq.heappop(max_heap)

            if max_heap:
                visited[max_heap[0][1]] = False
                heapq.heappop(max_heap)
              
        elif operation == 'D' and number == -1:
            while min_heap and not visited[min_heap[0][1]]:
                heapq.heappop(min_heap)
            
            if min_heap:
                visited[min_heap[0][1]] = False
                heapq.heappop(min_heap)

    while max_heap and not visited[max_heap[0][1]]:
        heapq.heappop(max_heap)      
    
    while min_heap and not visited[min_heap[0][1]]:
        heapq.heappop(min_heap)      
    
    if max_heap and min_heap:
        answer.append(-heapq.heappop(max_heap)[0])
        answer.append(heapq.heappop(min_heap)[0])
   
    else:
        answer.append(0)
        answer.append(0)
    
    return answer
